src/conversions/array.rs

```diff
@@ -0,0 +1,273 @@
+        // first try buffer protocol
+        if unsafe { crate::ffi::PyObject_CheckBuffer(obj.as_ptr()) } == 1 {
+            if let Ok(buf) = crate::buffer::PyBuffer::get(obj) {
+                if buf.dimensions() == 1 && buf.copy_to_slice(obj.py(), &mut array).is_ok() {
+                    buf.release(obj.py());
+                    // SAFETY: The array should be fully filled by `copy_to_slice`
+                    return Ok(unsafe { array.assume_init() });
+                }
+                buf.release(obj.py());
+            }
+        }
+        // fall back to sequence protocol
+        _extract_sequence_into_slice(obj, &mut array)?;
+        // SAFETY: The array should be fully filled by `_extract_sequence_into_slice`
+        Ok(unsafe { array.assume_init() })
+    }
+}
+
+#[cfg(min_const_generics)]
+fn create_array_from_obj<'s, T, const N: usize>(obj: &'s PyAny) -> PyResult<[T; N]>
+where
+    T: FromPyObject<'s>,
+{
+    let seq = <crate::types::PySequence as PyTryFrom>::try_from(obj)?;
+    let expected_len = seq.len()? as usize;
+    let mut counter = 0;
+    try_create_array(&mut counter, |idx| {
+        seq.get_item(idx as isize)
+            .map_err(|_| crate::utils::invalid_sequence_length(expected_len, idx + 1))?
+            .extract::<T>()
+    })
+}
+
+fn _extract_sequence_into_slice<'s, T>(obj: &'s PyAny, slice: &mut [T]) -> PyResult<()>
+where
+    T: FromPyObject<'s>,
+{
+    let seq = <crate::types::PySequence as PyTryFrom>::try_from(obj)?;
+    let expected_len = seq.len()? as usize;
+    if expected_len != slice.len() {
+        return Err(crate::utils::invalid_sequence_length(
+            expected_len,
+            slice.len(),
+        ));
+    }
+    for (value, item) in slice.iter_mut().zip(seq.iter()?) {
+        *value = item?.extract::<T>()?;
+    }
+    Ok(())
+}
+
+#[cfg(min_const_generics)]
+fn try_create_array<E, F, T, const N: usize>(counter: &mut usize, mut cb: F) -> Result<[T; N], E>
+where
+    F: FnMut(usize) -> Result<T, E>,
+{
+    // Helper to safely create arrays since the standard library doesn't
+    // provide one yet. Shouldn't be necessary in the future.
+    struct ArrayGuard<'a, T, const N: usize> {
+        dst: *mut T,
+        initialized: &'a mut usize,
+    }
+
+    impl<T, const N: usize> Drop for ArrayGuard<'_, T, N> {
+        fn drop(&mut self) {
+            debug_assert!(*self.initialized <= N);
+            let initialized_part = core::ptr::slice_from_raw_parts_mut(self.dst, *self.initialized);
+            unsafe {
+                core::ptr::drop_in_place(initialized_part);
+            }
+        }
+    }
+
+    let mut array: core::mem::MaybeUninit<[T; N]> = core::mem::MaybeUninit::uninit();
+    let guard: ArrayGuard<T, N> = ArrayGuard {
+        dst: array.as_mut_ptr() as _,
+        initialized: counter,
+    };
+    unsafe {
+        for (idx, value_ptr) in (&mut *array.as_mut_ptr()).iter_mut().enumerate() {
```

Comment:
I could be wrong; isn't this UB?

We're creating `value_ptr: &mut T` which refers to uninitialized memory; I'm reasonably certain that this breaks Rust's semantics of references and so the compiler is allowed to break this code at will.

I think the easiest way to avoid this could be to use `[MaybeUninit<T>; N]`.